### ISSUE TITLE
Enables the UserInfoService to extract OAuth2User from the additional parameters of the OAuth2AccessTokenResponse

### ISSUE THREAD

**Issue by XYUU** [spring-projects/spring-security#5370]
### Summary
Enables the UserInfoService to extract OAuth2User from the additional parameters of the OAuth2AccessTokenResponse. 
The user-info-uri is now not nullable, and some third-party OAuth2 service user information is already provided in the additional parameters of the OAuth2AccessTokenResponse.

### Expected Behavior
Some user information is returned directly in the AccessTokenResponse. Therefore, the userInfoUri is required to be empty. When the userInfoUri is empty, it is also possible to specify a DefaultOAuth2UserService.OAuth2UserExtractor so that he can extract the OAuth2User from the OAuth2UserRequest additionalParameters.

**Comment by jgrandja** [comment]
@XYUU I have not seen this use case as of yet:

> Some user information is returned directly in the `AccessTokenResponse.additionalParameters`

The `OAuth2UserService` is meant to retrieve the `UserInfo` from the **UserInfo Endpoint** using the Access Token returned from the `AccessTokenResponse.accessToken.tokenValue`.

However, if the returned `AccessTokenResponse.accessToken.tokenValue` is a JWT, than the `UserInfo` attributes may be extracted from the JWT in a custom implementation of [`OAuth2UserService`](https://docs.spring.io/spring-security/site/docs/5.0.5.RELEASE/reference/htmlsingle/#oauth2login-advanced-oauth2-user-service) as it receives it via `OAuth2UserRequest.OAuth2AccessToken.tokenValue`.

**Comment by XYUU** [comment]
See the OAuth service documentation provided by this OAuth 2.0 service provider：

This document is in Chinese. You can use Google to translate it into your desired language.

https://open.taobao.com/doc.htm?spm=a219a.7629065.1.21.Nh8kZB#?treeId=477&docId=102635&docType=1

Such as OAuth2 AccessToken Response:
```javascript
{
  "w2_expires_in": 0,
  "taobao_user_id": "263685215",
  "taobao_user_nick": "%E5%95%86%E5%AE%B6%E6%B5%8B%E8%AF%95%E5%B8%90%E5%8F%B752",
  "w1_expires_in": 1800,
  "re_expires_in": 0,
  "r2_expires_in": 0,
  "expires_in": 86400,
  "token_type": "Bearer",
  "refresh_token": "6200e1909ca29b04685c49d67f5ZZ3675347c0c6d5abccd263685215",
  "access_token": "6200819d9366af1383023a19907ZZf9048e4c14fd56333b263685215",
  "r1_expires_in": 1800
}
```

UserInfoService expects to extract：
```
  "taobao_user_id": "263685215",
  "taobao_user_nick": "%E5%95%86%E5%AE%B6%E6%B5%8B%E8%AF%95%E5%B8%90%E5%8F%B752",
```

**Comment by jgrandja** [comment]
@XYUU Could you not use

`"access_token": "6200819d9366af1383023a19907ZZf9048e4c14fd56333b263685215"`

to call the `UserInfo` endpoint exposed on the Provider, which in turn should return:

`"taobao_user_id": "263685215"` and other attributes

This is the purpose of `DefaultOAuth2UserService` or `OidcUserService`.

**Comment by jgrandja** [comment]
Looks like the User API is [here](https://open.taobao.com/docs/api_list.htm?spm=a219a.7629065.0.0.411iel&cid=1).

I'm not sure which API you need to configure for `spring.security.oauth2.provider.alipay.user-info-uri` but it should be in available with the Provider. The `DefaultOAuth2UserService` or `OidcUserService` will use the access token to call the configured UserInfo API endpoint to obtain the user information including `taobao_user_id`. This is the way the Client flow is designed. I believe you need to figure out which UserInfo Endpoint API to configure based on the exposed API's

**Comment by XYUU** [comment]
My understanding is as follows:

Taobao is a non-standard process OAuth2.0.
It does not provide UserInfo API.
Its user information is contained in AccessToken Response.
The above APIs have been verified. User information such as taobao_user_id and taobao_user_nick is sufficient.

So I think that in some non-standard processes, it may be allowed to not provide user-info-uri.

Actually I have implemented this feature. Do I need to commit a PR? Perhaps using code to explain it is easier to understand.

**Comment by jgrandja** [comment]
> Taobao is a non-standard process OAuth2.0.

I want to re-iterate again that Spring Security's goal is to provide spec-compliant implementations of OAuth 2.0 and OpenID Connect 1.0. If user's want/need to deviate from the spec than we will ensure there are extension points that they can hook into with their custom implementations.

The `OAuth2UserService` is meant to fetch user information from the provider. I haven't seen any provider as of yet that does not expose a "currently authenticated user" endpoint. I'm surprised that  Alipay does not. I would double check.

> Actually I have implemented this feature. Do I need to commit a PR?

I would prefer a working sample that demonstrates your use case. You can share the GitHub repo of your sample.

Please note: We will not change the core design of the OAuth 2.0 Client to adapt to a non-standard provider. If we can provide a hook that will suit your use case without changing the core design than we can consider this. It will be easier to understand from your sample.

**Comment by XYUU** [comment]
Yes, I agree with you. My needs also hope to provide extension points, but I think that the steps should be refined to make it easier to implement special extensions, rather than rewrite a lot of repetitive code. I believe you also agree with this goal.

Although Taobao and Alipay are the same company, the OAuth2.0 service they provide is two implementations. Do not confuse it. What I actually feed back is two different OAuth 2.0 adaptation issues.

**Comment by jgrandja** [comment]
@XYUU The additional parameters from `OAuth2AccessTokenResponse.getAdditionalParameters()` is now available via `OAuth2UserRequest.getAdditionalParameters()`.

As per your [comment](https://github.com/spring-projects/spring-security/issues/5370#issuecomment-393068350), you should now be able to access `taobao_user_id` and `taobao_user_nick` from `OAuth2UserRequest.getAdditionalParameters()`.

I'm going to close this issue as this is now available in Spring Security 5.1
